scipy/stats/stats.py

```diff
@@ -4275,6 +4277,158 @@ def mannwhitneyu(x, y, use_continuity=True):
+    test is to calculate the number of possible sequences for inputs of
+    length(x) and length(y), and to do an exact calculation if the number of
+    possible combinations is <100000. The default behavior may be overridden
+    with the use_exact flag.
+
+    This test corrects for ties and by default uses a continuity correction
+    when approximating the U statistic distribution.
+
+    The reported p-value is for a two-sided hypothesis. To get the one-sided
+    p-value set alternative to 'greater' or 'less' (default is 'two-sided').
+
+    For Mann-Whitney-Wilcoxon tests, the reported U statistic is the U used to
+    test the hypothesis. The u1 and u2 statistics are returned as well,
+    corresponding to x and y, respectively.
+
+
+    .. versionadded:: 0.17.0
+
+    References
+    ----------
+    .. [1] HB Mann and DR Whitney (1947). "On a Test of Whether one of Two
+            Random Variables is Stochastically Larger than the Other".
+            Annals of Mathematical Statistics. doi:10.1214/aoms/1177730491
+    .. [2] http://en.wikipedia.org/wiki/Mann-Whitney_U_test
+    """
+    x = asarray(x)
+    y = asarray(y)
+    n1 = len(x)
+    n2 = len(y)
+    if exact == 'auto':
+        exact = (n1 < 10 or n2 < 10) and n1 + n2 < 100000 \
+            and -np.log(n1 + n2 + 1) - special.betaln(n1 + 1, n2 + 1) < np.log(100000)
+    ranked = rankdata(np.concatenate((x,y)))
+    rankx = ranked[0:n1]       # get the x-ranks
+    T = tiecorrect(ranked)
+    if T == 0:
+        raise ValueError('All numbers are identical in mannwhitneyu')
+    if alternative not in ('two-sided', 'less', 'greater'):
+        raise AttributeError("Alternative should be one of: "
+                             "'two-sided', 'less', or 'greater'")
+    if exact:
+        a = list(range(n1, n1+n2))
+        u = [0]
+        while sum(a) != sum(range(n2)):   # When in leftmost position, a == list(range(n2))
+            # Do the shift operation
+            i = 0
+            while a[i] == i:
+                i += 1
+            a[i] -= 1
+            j = i - 1
+            while j >= 0:
+                a[j] = a[i] - (i - j)
+                j -= 1
+            # count(a < a1) = U2
+            u1 = 0
+            for i, x in enumerate(a):
+                u1 += n1 - x + i
+            u2 = n1*n2 - u1
+            # store min U value to array
+            if alternative == 'two-sided':
+                u.append(min(u1, u2))
+            else:
+                u.append(u1)
+        u1 = 0
+        u = np.array(u)
+        for i, x in enumerate(sorted(rankx)):
+            u1 += x - 1 - i
+        u2 = n1 * n2 - u1
+        if alternative == 'two-sided':
+            smallu = min(u1, u2)
+        elif alternative == 'greater':
+            smallu = u1
+        else:
+            smallu = u2
+        p = sum(u <= smallu) / len(u)
+        u = smallu
+    else:
+        u1 = n1*n2 + (n1*(n1+1))/2.0 - np.sum(rankx, axis=0)  # calc U for x
```

Comment:
I was wrong. After adding tests to verify that u1 and u2 are as expected, they are, in fact, reversed in the exact code (and this reversal was corrected in the subsequent testing). This has been fixed.